src/ui/pages/SetPasscode/SetPasscode.tsx

```diff
@@ -22,43 +22,108 @@ import { PageHeader } from "../../components/PageHeader";
+    "biometry.setupandroidbiometryheader"
+  );
+  const setupAndroidBiometryConfirmtext = i18n.t(
+    "biometry.setupandroidbiometryconfirm"
+  );
+  const setupAndroidBiometryCanceltext = i18n.t(
+    "biometry.setupandroidbiometrycancel"
+  );
+
+  const cancelBiometryHeaderText = i18n.t("biometry.cancelbiometryheader");
+  const cancelBiometryConfirmText = setupAndroidBiometryConfirmtext;
+
+  const handlePinChange = async (digit: number) => {
     if (passcode.length < 6) {
+      setPasscode(passcode + digit);
       if (originalPassCode !== "" && passcode.length === 5) {
         if (originalPassCode === passcode + digit) {
-          SecureStorage.set(KeyStoreKeys.APP_PASSCODE, originalPassCode).then(
-            () => {
-              const data: DataProps = {
-                store: { stateCache },
-              };
-              const { nextPath, updateRedux } = getNextRoute(
-                RoutePath.SET_PASSCODE,
-                data
-              );
-              updateReduxState(nextPath.pathname, data, dispatch, updateRedux);
-              ionRouter.push(nextPath.pathname, "forward", "push");
-              handleClearState();
-
-              PreferencesStorage.set(PreferencesKeys.APP_ALREADY_INIT, {
-                initialized: true,
-              }).then(() => dispatch(setInitialized(true)));
+          if (biometricInfo?.strongBiometryIsAvailable) {
+            if (isAndroidDevice) {
+              setShowSetupAndroidBiometryAlert(true);
+            } else {
+              await processBiometrics();
             }
-          );
+          } else {
+            await handlePassAuth();
+          }
+        }
+      }
+    }
+  };
+
+  const processBiometrics = async () => {
+    const isBiometricAuthenticated = await handleBiometricAuth();
+    if (isBiometricAuthenticated === true) {
+      await PreferencesStorage.set(PreferencesKeys.APP_BIOMETRY, {
+        enabled: true,
+      });
+      await handlePassAuth();
+    } else {
+      if (isBiometricAuthenticated instanceof BiometryError) {
+        if (
+          isBiometricAuthenticated.code === BiometryErrorType.userCancel ||
+          isBiometricAuthenticated.code ===
+            BiometryErrorType.biometryNotAvailable
+        ) {
+          setShowCancelBiometryAlert(true);
         }
       }
-      setPasscode(passcode + digit);
     }
   };
 
+  const handlePassAuth = async () => {
+    await SecureStorage.set(KeyStoreKeys.APP_PASSCODE, originalPassCode);
+    const data: DataProps = {
+      store: { stateCache },
+    };
+    const { nextPath, updateRedux } = getNextRoute(
+      RoutePath.SET_PASSCODE,
+      data
+    );
+    updateReduxState(nextPath.pathname, data, dispatch, updateRedux);
+    ionRouter.push(nextPath.pathname, "forward", "push");
+    handleClearState();
+
+    await PreferencesStorage.set(PreferencesKeys.APP_ALREADY_INIT, {
+      initialized: true,
+    });
+
+    dispatch(setInitialized(true));
```

Comment:
I don't think I reviewed this originally, but why is `setInitialized` here? In AppWrapper this is set after everything is "loaded" (KERIA can be down) so we stop showing the spinner.

Reply:
In AppWrapper we update the Redux state to indicate that the app has loaded all data before rendering. However, in this case, setInitialized is used to say that the app has completed the passcode creation process, marking its first use of the app. Then, the first step in the initial process of AppWrapper is to check if the app has been initialized; if not, we ensure that the keyStore is wiped with:
```
try {
      await PreferencesStorage.get(PreferencesKeys.APP_ALREADY_INIT);
    } catch (e) {
      await SecureStorage.delete(KeyStoreKeys.APP_PASSCODE);
      await SecureStorage.delete(KeyStoreKeys.IDENTITY_ENTROPY);
      await SecureStorage.delete(KeyStoreKeys.IDENTITY_ROOT_XPRV_KEY);
      await SecureStorage.delete(KeyStoreKeys.APP_OP_PASSWORD);
      await SecureStorage.delete(KeyStoreKeys.SIGNIFY_BRAN);
    }
```

Reply:
`APP_ALREADY_INIT` is different to `initialized` in Redux though, even if you set them both at the end of `SetPasscode.tsx`. I understand why you set `APP_ALREADY_INIT` (though this is temporary for dev purposes)

But `initialized` from what I can tell only seems to be used in `App.tsx` like this:
```
          {stateCache.initialized ? (
            <>
              {renderApp()}
              {!isPublicPage && !authentication.loggedIn ? <LockPage /> : null}
            </>
          ) : (
            <LoadingPage />
          )}
```
It indicates that the AppWrapper things have finished and we can stop showing the spinner and start showing the login screen. So I don't think it's necessary in `SetPasscode.tsx` - unless an error happened this will be `true` already anyway in the background from the AppWrapper.

Reply:
But also initialized is a different flag from the PreferencesStorage not Redux, so are awere that the user already used the app and it is not the first time.
Because in iOS, if the app is removed the keyStore persist so we need a that flag to wipe or not the KeyStore.

if this check fails:
`await PreferencesStorage.get(PreferencesKeys.APP_ALREADY_INIT);
`means that the user is using the app by the first time

Reply:
I am not talking about `APP_ALREADY_INIT` (which needs to go eventually).

I am talking about `stateCache.initialized` which is completely different. It is only used in `App.tsx` to decide on whether we show the spinner or not. This is only needed for AppWrapper initialisation, there is no need for it to be in this file.